### ISSUE TITLE
Rule "padded-blocks": add option to enforce padding in top-level declared functions

### ISSUE THREAD

**Issue by Monokai** [eslint/eslint#6578]
I'm using 3.0 and have this rule in my config:

```javascript
"padded-blocks": ["error", {
  "blocks": "never",
  "classes": "always"
}]
```

This correctly enforces a formatting like this:

```javascript
export default class {

  constructor() {
  }

  a() {
  }

  b() {
  }

}
```

Sometimes I'd like to export a module which is not a class, but is an IIFE (and has a revealing module pattern): 

```javascript
export default (function() {

  function a() {
  }

  function b() {
  }

  return {
    b
  }

})();
```

Essentially this has the same structure as an ES6 defined class, but I cannot configure `padded-blocks` to allow this formatting. It would be nice if we have a separate option for (exported) top-level declared functions.

**Comment by vitorbal** [comment]
Hi @Monokai! Thanks for your suggestion. I can see the value in this but I wonder if we shouldn't generalize it to all IIFE's, since the `classes` option also applies to all classes and not only classes that are being exported. What do you think?

Also, would you be willing to work on a PR for this if it gets accepted?

**Comment by Monokai** [comment]
@vitorbal I'm not sure. You can also have something like this IIFE for scoping in the middle of your code (admittedly not needed when using ES6 arrow functions):

```javascript
for (let i = 0; i < elements.length; i++) {
	(function(index) {
		elements[i].addEventListener('click', function() {
			console.log(index);
		});
	})(i);
}
```

I can see some options:

1. Make a rule for top-level functions
  This also allows class-like function declarations for those that don't like to use classes but regular functions instead.

2. Make a rule for IIFE's
  This will also force padding on IIFE's in the middle of your code.

3. Make a rule for exported IIFE's
   This wouldn't allow regular function declarations that act as classes.

4. Make a rule for revealing module pattern style functions
  This would allow IIFE's and regular functions. But how to define this? A function that contains methods and returns an object referencing these methods? How about a module that has no methods?

This is harder than I initially thought. I think the top-level function rule is best. I'd have to find time and dive into the docs to see how I would make a PR for this, but first we need consensus I guess. What do you think?

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
